### PR TITLE
Trainer see

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: include/constants/event_object_movement.h [hunk 1]
```diff
@@ -146,100 +146,102 @@
...
+#define MOVEMENT_ACTION_STORE_AND_LOCK_ANIM             0xA2
+#define MOVEMENT_ACTION_FREE_AND_UNLOCK_ANIM            0xA3
+#define MOVEMENT_ACTION_WALK_LEFT_AFFINE                0xA4
+#define MOVEMENT_ACTION_WALK_RIGHT_AFFINE               0xA5
+#define MOVEMENT_ACTION_LEVITATE                        0xA6
+#define MOVEMENT_ACTION_STOP_LEVITATE                   0xA7
+#define MOVEMENT_ACTION_DESTROY_EXTRA_TASK_IF_AT_TOP    0xA8
+#define MOVEMENT_ACTION_FIGURE_8                        0xA9
 
 // These two are verified correct in FRLG
```
Comment (jiangzhengwenjz) [inline include/constants/event_object_movement.h]: Do they conflict with above?

File: src/trainer_see.c [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+    return FALSE;
+}
+
+static bool8 TrainerSeeFunc_WaitJumpOutOfAsh(u8 taskId, struct Task * task, struct ObjectEvent * trainerObj)
+{
+    struct Sprite *sprite;
+
+    if (gSprites[task->tOutOfAshSpriteId].animCmdIndex == 2)
+    {
+        trainerObj->fixedPriority = 0;
```
Comment (jiangzhengwenjz) [inline src/trainer_see.c]: `FALSE`

File: src/trainer_see.c [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+        trainerObj->triggerGroundEffectsOnMove = 1;
```
Comment (jiangzhengwenjz) [inline src/trainer_see.c]: `TRUE`

File: src/trainer_see.c [hunk 4]
```diff
@@ -0,0 +1,753 @@
...
+    sTrainerSeeFuncList2[task->data[0]](taskId, task, trainerObj);
+    if (task->data[0] == 3 && !FieldEffectActiveListContains(FLDEFF_POP_OUT_OF_ASH))
+    {
+        SetTrainerMovementType(trainerObj, GetTrainerFacingDirectionMovementType(trainerObj->facingDirection));
+        TryOverrideTemplateCoordsForObjectEvent(trainerObj, GetTrainerFacingDirectionMovementType(trainerObj->facingDirection));
+        DestroyTask(taskId);
+    }
+    else
+    {
+        trainerObj->heldMovementFinished = 0;
```
Comment (jiangzhengwenjz) [inline src/trainer_see.c]: `FALSE`

File: src/trainer_see.c [hunk 5]
```diff
@@ -0,0 +1,753 @@
...
+    if (spriteId != MAX_SPRITES)
+        SetIconSpriteData(&gSprites[spriteId], FLDEFF_EXCLAMATION_MARK_ICON_2, 4);
+
+    return 0;
+}
+
+static void SetIconSpriteData(struct Sprite *sprite, u16 fldEffId, u8 spriteAnimNum)
+{
+    sprite->oam.priority = 1;
+    sprite->coordOffsetEnabled = 1;
```
Comment (jiangzhengwenjz) [inline src/trainer_see.c]: `TRUE`

File: src/trainer_see.c [hunk 6]
```diff
@@ -0,0 +1,753 @@
...
+#include "task.h"
+#include "trainer_see.h"
+#include "util.h"
+#include "constants/battle_setup.h"
+#include "constants/event_object_movement.h"
+#include "constants/event_objects.h"
+#include "constants/object_events.h"
+
+typedef u8 (*TRAINER_APPROACH_FUNC)(struct ObjectEvent *, s16, s16, s16);
+typedef bool8 (*TRAINER_SEE_FUNC)(u8, struct Task *, struct ObjectEvent *);
```
Comment (jiangzhengwenjz) [inline src/trainer_see.c]: I think other custom types are in pascal case except libs?

File: include/constants/event_object_movement.h (hunk 1 above)
Comment (PikalaxALT) [reply include/constants/event_object_movement.h]: Unknown, to be determined

File: data/maps/CeladonCity_GameCorner/scripts.inc [hunk 7]
```diff
@@ -257,8 +257,8 @@ CeladonCity_GameCorner_EventScript_SlotMachine:: @ 816C96C
 	setfieldeffectarg 0, 255
 	setfieldeffectarg 1, 10
 	setfieldeffectarg 2, 14
-	dofieldeffect FLDEFF_UNK_40
-	waitfieldeffect FLDEFF_UNK_40  @ Smile icon?
+	dofieldeffect FLDEFF_SMILEY_FACE_ICON
+	waitfieldeffect FLDEFF_SMILEY_FACE_ICON  @ Smile icon?
```
Comment (GriffinRichards) [inline data/maps/CeladonCity_GameCorner/scripts.inc]: Comment no longer needed

File: include/constants/field_effects.h [hunk 8]
```diff
@@ -65,9 +65,9 @@
 #define FLDEFF_PCTURN_ON                 61
 #define FLDEFF_HALL_OF_FAME_RECORD       62
 #define FLDEFF_USE_TELEPORT              63
-#define FLDEFF_UNK_40                    64 // TODO: game corner smile icon?
+#define FLDEFF_SMILEY_FACE_ICON                    64 // TODO: game corner smile icon?
 #define FLDEFF_UNK_41                    65 // TODO: related to vs_seeker
-#define FLDEFF_UNK_42                    66 // TODO: related to trainer see
+#define FLDEFF_DOUBLE_EXCL_MARK_ICON                    66 // TODO: related to trainer see
```
Comment (GriffinRichards) [inline include/constants/field_effects.h]: Comments no longer needed, and re-align

File: src/rfu_union_tool.c [hunk 9]
```diff
@@ -1,7 +1,7 @@
 #include "global.h"
 #include "event_data.h"
 #include "event_object_movement.h"
-#include "event_object_80688E4.h"
+#include "event_object_movement.h"
```
Comment (jiangzhengwenjz) [inline src/rfu_union_tool.c]: duplicate

File: src/quest_log.c [hunk 10]
```diff
@@ -18,7 +18,7 @@
 #include "field_fadetransition.h"
 #include "field_weather.h"
 #include "event_object_movement.h"
-#include "event_object_80688E4.h"
+#include "event_object_movement.h"
```
Comment (jiangzhengwenjz) [inline src/quest_log.c]: duplicate

File: src/event_object_lock.c [hunk 11]
```diff
@@ -3,7 +3,7 @@
 #include "field_player_avatar.h"
 #include "event_object_movement.h"
 #include "script_movement.h"
-#include "event_object_80688E4.h"
+#include "event_object_movement.h"
```
Comment (jiangzhengwenjz) [inline src/event_object_lock.c]: duplicate

File: src/event_object_80688E4.c [hunk 12]
```diff
@@ -1,5 +1,5 @@
 #include "global.h"
-#include "event_object_80688E4.h"
+#include "event_object_movement.h"
 #include "event_object_movement.h"
```
Comment (jiangzhengwenjz) [inline src/event_object_80688E4.c]: duplicate

File: src/vs_seeker.c [hunk 13]
```diff
@@ -20,7 +20,7 @@
 #include "random.h"
 #include "event_object_movement.h"
 #include "field_player_avatar.h"
-#include "event_object_80688E4.h"
+#include "event_object_movement.h"
```
Comment (jiangzhengwenjz) [inline src/vs_seeker.c]: duplicate

File: include/constants/event_object_movement.h [hunk 14]
```diff
@@ -126,10 +133,10 @@
...
 #define MOVEMENT_ACTION_WALK_IN_PLACE_FASTEST_LEFT      0x2F
 #define MOVEMENT_ACTION_WALK_IN_PLACE_FASTEST_RIGHT     0x30
-#define MOVEMENT_ACTION_RIDE_WATER_CURRENT_DOWN         0x31
-#define MOVEMENT_ACTION_RIDE_WATER_CURRENT_UP           0x32
-#define MOVEMENT_ACTION_RIDE_WATER_CURRENT_LEFT         0x33
-#define MOVEMENT_ACTION_RIDE_WATER_CURRENT_RIGHT        0x34
+#define MOVEMENT_ACTION_FASE_DOWN_SLOW                  0x31
+#define MOVEMENT_ACTION_FASE_UP_SLOW                    0x32
+#define MOVEMENT_ACTION_FASE_LEFT_SLOW                  0x33
+#define MOVEMENT_ACTION_FASE_RIGHT_SLOW                 0x34
```
Comment (GriffinRichards) [inline include/constants/event_object_movement.h]: `FACE`

**Pull request merged**